{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare the input data objects\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```micromamba activate gena-env```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import multianndata as mad\n",
    "import cna\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# constants\n",
    "outdir = (\n",
    "    \"/directflow/SCCGGroupShare/projects/blabow/tenk10k_phase1/data_processing/csa_qtl\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. prepare the required inputs:\n",
    "    * cell metadata with donor info for each cell \n",
    "    * cell by gene counts matrix \n",
    "    * donor-level metadata storing covariate information to be used in the GWAS, as well as batches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the latest tenk cohort\n",
    "adata = sc.read(\n",
    "    \"/directflow/SCCGGroupShare/projects/blabow/tenk10k_phase1/data_processing/scanpy/output/integrated_objects/240_libraries/240_libraries_concatenated_harmony_leiden_filtered_reanalysed.h5ad\",\n",
    "    backed='r'\n",
    ")\n",
    "# NOTE: reading in backed mode is 100x faster although I can't modify the object \n",
    "# NOTE: can also try out \"caching\" the h5ad for faster read times as well??\n",
    "\n",
    "## Currently we have only TOB genotypes available on brenner\n",
    "# filter for just TOB to do a practice run\n",
    "adata = adata[adata.obs[\"cohort\"] == \"TOB\"]\n",
    "### ^^ Remove this when I want to do analysis for Bioheart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellbender_background_fraction</th>\n",
       "      <th>cellbender_cell_probability</th>\n",
       "      <th>cellbender_cell_size</th>\n",
       "      <th>cellbender_droplet_efficiency</th>\n",
       "      <th>celltypist_predicted_labels</th>\n",
       "      <th>celltypist_over_clustering</th>\n",
       "      <th>celltypist_majority_voting</th>\n",
       "      <th>celltypist_conf_score</th>\n",
       "      <th>wg2_sample</th>\n",
       "      <th>wg2_nCount_RNA</th>\n",
       "      <th>wg2_nFeature_RNA</th>\n",
       "      <th>wg2_percent_mt</th>\n",
       "      <th>wg2_azimuth_predicted_celltype_l2</th>\n",
       "      <th>wg2_azimuth_predicted_celltype_l2_score</th>\n",
       "      <th>wg2_scpred_prediction</th>\n",
       "      <th>Vireo_Individual_Assignment</th>\n",
       "      <th>Vireo_DropletType</th>\n",
       "      <th>scDblFinder_DropletType</th>\n",
       "      <th>scDblFinder_Score</th>\n",
       "      <th>scds_score</th>\n",
       "      <th>scds_DropletType</th>\n",
       "      <th>MajoritySinglet_DropletType</th>\n",
       "      <th>MajoritySinglet_Individual_Assignment</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>total_counts_ribo</th>\n",
       "      <th>pct_counts_ribo</th>\n",
       "      <th>total_counts_hb</th>\n",
       "      <th>pct_counts_hb</th>\n",
       "      <th>original_barcode</th>\n",
       "      <th>new_cell_name</th>\n",
       "      <th>sequencing_library</th>\n",
       "      <th>individual</th>\n",
       "      <th>cohort</th>\n",
       "      <th>onek1k_id</th>\n",
       "      <th>cpg_id_old</th>\n",
       "      <th>tob_id</th>\n",
       "      <th>cpg_id</th>\n",
       "      <th>onek1k_donor</th>\n",
       "      <th>ct_id</th>\n",
       "      <th>batch</th>\n",
       "      <th>leiden</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATCGAT_S0012b</th>\n",
       "      <td>0.009700</td>\n",
       "      <td>0.999954</td>\n",
       "      <td>7748.034668</td>\n",
       "      <td>1.187165</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>4</td>\n",
       "      <td>Tcm/Naive helper T cells</td>\n",
       "      <td>0.988154</td>\n",
       "      <td>S0012b</td>\n",
       "      <td>8454.0</td>\n",
       "      <td>2621.0</td>\n",
       "      <td>2.247457</td>\n",
       "      <td>CD4_TCM</td>\n",
       "      <td>0.681302</td>\n",
       "      <td>CD4_Naive</td>\n",
       "      <td>558</td>\n",
       "      <td>singlet</td>\n",
       "      <td>singlet</td>\n",
       "      <td>0.007869</td>\n",
       "      <td>0.279380</td>\n",
       "      <td>singlet</td>\n",
       "      <td>singlet</td>\n",
       "      <td>558</td>\n",
       "      <td>2621</td>\n",
       "      <td>2621</td>\n",
       "      <td>8454.0</td>\n",
       "      <td>190.0</td>\n",
       "      <td>2.247457</td>\n",
       "      <td>2676.0</td>\n",
       "      <td>31.653654</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.023657</td>\n",
       "      <td>AAACCCAAGAATCGAT-1</td>\n",
       "      <td>AAACCCAAGAATCGAT_S0012b</td>\n",
       "      <td>S0012b</td>\n",
       "      <td>CPG313965</td>\n",
       "      <td>TOB</td>\n",
       "      <td>557_558</td>\n",
       "      <td>CPG6106</td>\n",
       "      <td>TOB1413</td>\n",
       "      <td>CPG313965</td>\n",
       "      <td>558</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAGCAACC_S0012b</th>\n",
       "      <td>0.018032</td>\n",
       "      <td>0.999963</td>\n",
       "      <td>7818.318848</td>\n",
       "      <td>0.806677</td>\n",
       "      <td>Tem/Temra cytotoxic T cells</td>\n",
       "      <td>3</td>\n",
       "      <td>Tem/Temra cytotoxic T cells</td>\n",
       "      <td>0.339064</td>\n",
       "      <td>S0012b</td>\n",
       "      <td>5823.0</td>\n",
       "      <td>2328.0</td>\n",
       "      <td>5.667182</td>\n",
       "      <td>CD4_CTL</td>\n",
       "      <td>0.616923</td>\n",
       "      <td>CD4_CTL</td>\n",
       "      <td>27</td>\n",
       "      <td>singlet</td>\n",
       "      <td>doublet</td>\n",
       "      <td>0.782162</td>\n",
       "      <td>0.102471</td>\n",
       "      <td>singlet</td>\n",
       "      <td>singlet</td>\n",
       "      <td>27</td>\n",
       "      <td>2328</td>\n",
       "      <td>2328</td>\n",
       "      <td>5823.0</td>\n",
       "      <td>330.0</td>\n",
       "      <td>5.667181</td>\n",
       "      <td>957.0</td>\n",
       "      <td>16.434828</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AAACCCAAGAGCAACC-1</td>\n",
       "      <td>AAACCCAAGAGCAACC_S0012b</td>\n",
       "      <td>S0012b</td>\n",
       "      <td>CPG309211</td>\n",
       "      <td>TOB</td>\n",
       "      <td>27_27</td>\n",
       "      <td>CPG943</td>\n",
       "      <td>TOB1642</td>\n",
       "      <td>CPG309211</td>\n",
       "      <td>27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGACAGT_S0012b</th>\n",
       "      <td>0.262001</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>4126.960449</td>\n",
       "      <td>0.783668</td>\n",
       "      <td>Tem/Trm cytotoxic T cells</td>\n",
       "      <td>49</td>\n",
       "      <td>Tem/Temra cytotoxic T cells</td>\n",
       "      <td>0.011127</td>\n",
       "      <td>S0012b</td>\n",
       "      <td>2729.0</td>\n",
       "      <td>1519.0</td>\n",
       "      <td>14.254306</td>\n",
       "      <td>CD8_TEM</td>\n",
       "      <td>0.461040</td>\n",
       "      <td>CD8_TEM</td>\n",
       "      <td>502</td>\n",
       "      <td>singlet</td>\n",
       "      <td>singlet</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.094898</td>\n",
       "      <td>singlet</td>\n",
       "      <td>singlet</td>\n",
       "      <td>502</td>\n",
       "      <td>1519</td>\n",
       "      <td>1519</td>\n",
       "      <td>2729.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>14.254306</td>\n",
       "      <td>84.0</td>\n",
       "      <td>3.078051</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.036643</td>\n",
       "      <td>AAACCCAAGCGACAGT-1</td>\n",
       "      <td>AAACCCAAGCGACAGT_S0012b</td>\n",
       "      <td>S0012b</td>\n",
       "      <td>CPG313437</td>\n",
       "      <td>TOB</td>\n",
       "      <td>501_502</td>\n",
       "      <td>CPG5546</td>\n",
       "      <td>TOB1337</td>\n",
       "      <td>CPG313437</td>\n",
       "      <td>502</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         cellbender_background_fraction  \\\n",
       "AAACCCAAGAATCGAT_S0012b                        0.009700   \n",
       "AAACCCAAGAGCAACC_S0012b                        0.018032   \n",
       "AAACCCAAGCGACAGT_S0012b                        0.262001   \n",
       "\n",
       "                         cellbender_cell_probability  cellbender_cell_size  \\\n",
       "AAACCCAAGAATCGAT_S0012b                     0.999954           7748.034668   \n",
       "AAACCCAAGAGCAACC_S0012b                     0.999963           7818.318848   \n",
       "AAACCCAAGCGACAGT_S0012b                     0.999995           4126.960449   \n",
       "\n",
       "                         cellbender_droplet_efficiency  \\\n",
       "AAACCCAAGAATCGAT_S0012b                       1.187165   \n",
       "AAACCCAAGAGCAACC_S0012b                       0.806677   \n",
       "AAACCCAAGCGACAGT_S0012b                       0.783668   \n",
       "\n",
       "                         celltypist_predicted_labels  \\\n",
       "AAACCCAAGAATCGAT_S0012b     Tcm/Naive helper T cells   \n",
       "AAACCCAAGAGCAACC_S0012b  Tem/Temra cytotoxic T cells   \n",
       "AAACCCAAGCGACAGT_S0012b    Tem/Trm cytotoxic T cells   \n",
       "\n",
       "                        celltypist_over_clustering  \\\n",
       "AAACCCAAGAATCGAT_S0012b                          4   \n",
       "AAACCCAAGAGCAACC_S0012b                          3   \n",
       "AAACCCAAGCGACAGT_S0012b                         49   \n",
       "\n",
       "                          celltypist_majority_voting  celltypist_conf_score  \\\n",
       "AAACCCAAGAATCGAT_S0012b     Tcm/Naive helper T cells               0.988154   \n",
       "AAACCCAAGAGCAACC_S0012b  Tem/Temra cytotoxic T cells               0.339064   \n",
       "AAACCCAAGCGACAGT_S0012b  Tem/Temra cytotoxic T cells               0.011127   \n",
       "\n",
       "                        wg2_sample  wg2_nCount_RNA  wg2_nFeature_RNA  \\\n",
       "AAACCCAAGAATCGAT_S0012b     S0012b          8454.0            2621.0   \n",
       "AAACCCAAGAGCAACC_S0012b     S0012b          5823.0            2328.0   \n",
       "AAACCCAAGCGACAGT_S0012b     S0012b          2729.0            1519.0   \n",
       "\n",
       "                         wg2_percent_mt wg2_azimuth_predicted_celltype_l2  \\\n",
       "AAACCCAAGAATCGAT_S0012b        2.247457                           CD4_TCM   \n",
       "AAACCCAAGAGCAACC_S0012b        5.667182                           CD4_CTL   \n",
       "AAACCCAAGCGACAGT_S0012b       14.254306                           CD8_TEM   \n",
       "\n",
       "                         wg2_azimuth_predicted_celltype_l2_score  \\\n",
       "AAACCCAAGAATCGAT_S0012b                                 0.681302   \n",
       "AAACCCAAGAGCAACC_S0012b                                 0.616923   \n",
       "AAACCCAAGCGACAGT_S0012b                                 0.461040   \n",
       "\n",
       "                        wg2_scpred_prediction Vireo_Individual_Assignment  \\\n",
       "AAACCCAAGAATCGAT_S0012b             CD4_Naive                         558   \n",
       "AAACCCAAGAGCAACC_S0012b               CD4_CTL                          27   \n",
       "AAACCCAAGCGACAGT_S0012b               CD8_TEM                         502   \n",
       "\n",
       "                        Vireo_DropletType scDblFinder_DropletType  \\\n",
       "AAACCCAAGAATCGAT_S0012b           singlet                 singlet   \n",
       "AAACCCAAGAGCAACC_S0012b           singlet                 doublet   \n",
       "AAACCCAAGCGACAGT_S0012b           singlet                 singlet   \n",
       "\n",
       "                         scDblFinder_Score  scds_score scds_DropletType  \\\n",
       "AAACCCAAGAATCGAT_S0012b           0.007869    0.279380          singlet   \n",
       "AAACCCAAGAGCAACC_S0012b           0.782162    0.102471          singlet   \n",
       "AAACCCAAGCGACAGT_S0012b           0.000019    0.094898          singlet   \n",
       "\n",
       "                        MajoritySinglet_DropletType  \\\n",
       "AAACCCAAGAATCGAT_S0012b                     singlet   \n",
       "AAACCCAAGAGCAACC_S0012b                     singlet   \n",
       "AAACCCAAGCGACAGT_S0012b                     singlet   \n",
       "\n",
       "                        MajoritySinglet_Individual_Assignment  n_genes  \\\n",
       "AAACCCAAGAATCGAT_S0012b                                   558     2621   \n",
       "AAACCCAAGAGCAACC_S0012b                                    27     2328   \n",
       "AAACCCAAGCGACAGT_S0012b                                   502     1519   \n",
       "\n",
       "                         n_genes_by_counts  total_counts  total_counts_mt  \\\n",
       "AAACCCAAGAATCGAT_S0012b               2621        8454.0            190.0   \n",
       "AAACCCAAGAGCAACC_S0012b               2328        5823.0            330.0   \n",
       "AAACCCAAGCGACAGT_S0012b               1519        2729.0            389.0   \n",
       "\n",
       "                         pct_counts_mt  total_counts_ribo  pct_counts_ribo  \\\n",
       "AAACCCAAGAATCGAT_S0012b       2.247457             2676.0        31.653654   \n",
       "AAACCCAAGAGCAACC_S0012b       5.667181              957.0        16.434828   \n",
       "AAACCCAAGCGACAGT_S0012b      14.254306               84.0         3.078051   \n",
       "\n",
       "                         total_counts_hb  pct_counts_hb    original_barcode  \\\n",
       "AAACCCAAGAATCGAT_S0012b              2.0       0.023657  AAACCCAAGAATCGAT-1   \n",
       "AAACCCAAGAGCAACC_S0012b              0.0       0.000000  AAACCCAAGAGCAACC-1   \n",
       "AAACCCAAGCGACAGT_S0012b              1.0       0.036643  AAACCCAAGCGACAGT-1   \n",
       "\n",
       "                                   new_cell_name sequencing_library  \\\n",
       "AAACCCAAGAATCGAT_S0012b  AAACCCAAGAATCGAT_S0012b             S0012b   \n",
       "AAACCCAAGAGCAACC_S0012b  AAACCCAAGAGCAACC_S0012b             S0012b   \n",
       "AAACCCAAGCGACAGT_S0012b  AAACCCAAGCGACAGT_S0012b             S0012b   \n",
       "\n",
       "                        individual cohort onek1k_id cpg_id_old   tob_id  \\\n",
       "AAACCCAAGAATCGAT_S0012b  CPG313965    TOB   557_558    CPG6106  TOB1413   \n",
       "AAACCCAAGAGCAACC_S0012b  CPG309211    TOB     27_27     CPG943  TOB1642   \n",
       "AAACCCAAGCGACAGT_S0012b  CPG313437    TOB   501_502    CPG5546  TOB1337   \n",
       "\n",
       "                            cpg_id onek1k_donor ct_id batch leiden  \n",
       "AAACCCAAGAATCGAT_S0012b  CPG313965          558   NaN     0      6  \n",
       "AAACCCAAGAGCAACC_S0012b  CPG309211           27   NaN     0      2  \n",
       "AAACCCAAGCGACAGT_S0012b  CPG313437          502   NaN     0      2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context('display.max_columns', None):\n",
    "    display(adata.obs.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: looks like there are 42 TOB samples with CPG ids without onek1k study id's. Possible that they have no WGS and therefore no CPG ids, or they are somehow missing for the master metadata sheet. I will just remove these for now..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATCGAT_S0012b</th>\n",
       "      <td>557_558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAGCAACC_S0012b</th>\n",
       "      <td>27_27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGACAGT_S0012b</th>\n",
       "      <td>501_502</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              id\n",
       "AAACCCAAGAATCGAT_S0012b  557_558\n",
       "AAACCCAAGAGCAACC_S0012b    27_27\n",
       "AAACCCAAGCGACAGT_S0012b  501_502"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_meta = adata.obs[[\"onek1k_id\"]].set_index(adata.obs.index).rename(columns={\"onek1k_id\": \"id\"})\n",
    "cell_meta.to_csv(f\"{outdir}/data/cell_meta.csv\")\n",
    "display(cell_meta.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequencing_library</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>onek1k_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>557_558</th>\n",
       "      <td>S0012b</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27_27</th>\n",
       "      <td>S0012b</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501_502</th>\n",
       "      <td>S0012b</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sequencing_library  batch\n",
       "onek1k_id                          \n",
       "557_558               S0012b      0\n",
       "27_27                 S0012b      0\n",
       "501_502               S0012b      0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>onek1k_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>557_558</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27_27</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501_502</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491_492</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500_501</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391_392</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348_349</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330_331</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049_1050</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_1054</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3735 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           batch\n",
       "onek1k_id       \n",
       "557_558        0\n",
       "27_27          0\n",
       "501_502        0\n",
       "491_492        0\n",
       "500_501        0\n",
       "...          ...\n",
       "391_392      119\n",
       "348_349      119\n",
       "330_331      119\n",
       "1049_1050    119\n",
       "1053_1054    119\n",
       "\n",
       "[3735 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get seq library to cpg id mapping - use this to remove batchy neighbourhoods later on\n",
    "sequencing_library_mapping = adata.obs[\n",
    "    [\"onek1k_id\", \"sequencing_library\"]\n",
    "].drop_duplicates()\n",
    "sequencing_library_mapping = sequencing_library_mapping.set_index(\n",
    "    \"onek1k_id\", drop=True\n",
    ")\n",
    "sequencing_library_mapping[\"batch\"] = sequencing_library_mapping[\n",
    "    \"sequencing_library\"\n",
    "].factorize()[0]\n",
    "display(sequencing_library_mapping.head(3))\n",
    "batch_mapping = sequencing_library_mapping.drop(\"sequencing_library\", axis=1)\n",
    "display(batch_mapping)\n",
    "sequencing_library_mapping.to_csv(f\"{outdir}/data/sequencing_library_mapping.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " NOTE:\n",
    " * Figure out how to add \"batch\" to samplem and remove batchy neighbourhoods when running cna \n",
    " * currently this is not possible / can't figure out how to do it when I have a many:many mapping between batches and samples (onek1k id's)?\n",
    " * this just makes the samplem have multiple rows per sample which feels wrong \n",
    " * can maybe find an answer to this in \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequencing_library</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>onek1k_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>S0026c</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>S0026a</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>S0026b</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>S0025_28b</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>S0025_28a</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>S0026b</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>S0026a</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>S0026c</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>S0025_28a</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_2</th>\n",
       "      <td>S0025_28b</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>S0032a</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>S0030_33a</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>S0030_33b</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>S0032b</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_3</th>\n",
       "      <td>S0032c</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>S0030_33b</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>S0030_33a</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>S0032a</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>S0032b</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_4</th>\n",
       "      <td>S0032c</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>S0018b</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>S0018c</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_6</th>\n",
       "      <td>S0018a</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7_7</th>\n",
       "      <td>S0018a</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7_7</th>\n",
       "      <td>S0018b</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7_7</th>\n",
       "      <td>S0018c</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8_8</th>\n",
       "      <td>S0018b</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8_8</th>\n",
       "      <td>S0018a</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8_8</th>\n",
       "      <td>S0018c</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_9</th>\n",
       "      <td>S0025_28b</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_9</th>\n",
       "      <td>S0025_28a</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_9</th>\n",
       "      <td>S0028b</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_9</th>\n",
       "      <td>S0028c</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_9</th>\n",
       "      <td>S0028a</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10</th>\n",
       "      <td>S0028c</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10</th>\n",
       "      <td>S0025_28b</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10</th>\n",
       "      <td>S0028a</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10</th>\n",
       "      <td>S0025_28a</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10</th>\n",
       "      <td>S0028b</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11_11</th>\n",
       "      <td>S0028a</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11_11</th>\n",
       "      <td>S0025_28b</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11_11</th>\n",
       "      <td>S0028b</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11_11</th>\n",
       "      <td>S0028c</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11_11</th>\n",
       "      <td>S0025_28a</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12_12</th>\n",
       "      <td>S0017b</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12_12</th>\n",
       "      <td>S0017c</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13_13</th>\n",
       "      <td>S0017c</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13_13</th>\n",
       "      <td>S0017b</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15_15</th>\n",
       "      <td>S0022a</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15_15</th>\n",
       "      <td>S0021_24a</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15_15</th>\n",
       "      <td>S0022c</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15_15</th>\n",
       "      <td>S0021_24b</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15_15</th>\n",
       "      <td>S0022b</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16_16</th>\n",
       "      <td>S0022c</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16_16</th>\n",
       "      <td>S0022b</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16_16</th>\n",
       "      <td>S0021_24a</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16_16</th>\n",
       "      <td>S0021_24b</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16_16</th>\n",
       "      <td>S0022a</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17_17</th>\n",
       "      <td>S0022c</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17_17</th>\n",
       "      <td>S0022a</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17_17</th>\n",
       "      <td>S0022b</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17_17</th>\n",
       "      <td>S0021_24b</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17_17</th>\n",
       "      <td>S0021_24a</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18_18</th>\n",
       "      <td>S0021_24a</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18_18</th>\n",
       "      <td>S0022c</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18_18</th>\n",
       "      <td>S0022a</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18_18</th>\n",
       "      <td>S0021_24b</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18_18</th>\n",
       "      <td>S0022b</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19_19</th>\n",
       "      <td>S0021_24b</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19_19</th>\n",
       "      <td>S0022c</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19_19</th>\n",
       "      <td>S0022b</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19_19</th>\n",
       "      <td>S0021_24a</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19_19</th>\n",
       "      <td>S0022a</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23_23</th>\n",
       "      <td>S0022a</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23_23</th>\n",
       "      <td>S0021_24a</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23_23</th>\n",
       "      <td>S0022b</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23_23</th>\n",
       "      <td>S0022c</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23_23</th>\n",
       "      <td>S0021_24b</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24_24</th>\n",
       "      <td>S0020b</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24_24</th>\n",
       "      <td>S0020c</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24_24</th>\n",
       "      <td>S0020a</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25_25</th>\n",
       "      <td>S0012c</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25_25</th>\n",
       "      <td>S0012b</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25_25</th>\n",
       "      <td>S0012a</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27_27</th>\n",
       "      <td>S0012c</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27_27</th>\n",
       "      <td>S0012a</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27_27</th>\n",
       "      <td>S0012b</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28_28</th>\n",
       "      <td>S0012c</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28_28</th>\n",
       "      <td>S0012a</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28_28</th>\n",
       "      <td>S0012b</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29_29</th>\n",
       "      <td>S0025_28a</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29_29</th>\n",
       "      <td>S0026a</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29_29</th>\n",
       "      <td>S0026c</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29_29</th>\n",
       "      <td>S0026b</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29_29</th>\n",
       "      <td>S0025_28b</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30_30</th>\n",
       "      <td>S0025_28a</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30_30</th>\n",
       "      <td>S0026a</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30_30</th>\n",
       "      <td>S0026b</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30_30</th>\n",
       "      <td>S0026c</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30_30</th>\n",
       "      <td>S0025_28b</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sequencing_library  batch\n",
       "onek1k_id                          \n",
       "1_1                   S0026c     56\n",
       "1_1                   S0026a     61\n",
       "1_1                   S0026b     48\n",
       "1_1                S0025_28b     59\n",
       "1_1                S0025_28a     52\n",
       "2_2                   S0026b     48\n",
       "2_2                   S0026a     61\n",
       "2_2                   S0026c     56\n",
       "2_2                S0025_28a     52\n",
       "2_2                S0025_28b     59\n",
       "3_3                   S0032a     51\n",
       "3_3                S0030_33a     33\n",
       "3_3                S0030_33b     98\n",
       "3_3                   S0032b     39\n",
       "3_3                   S0032c     49\n",
       "4_4                S0030_33b     98\n",
       "4_4                S0030_33a     33\n",
       "4_4                   S0032a     51\n",
       "4_4                   S0032b     39\n",
       "4_4                   S0032c     49\n",
       "6_6                   S0018b     58\n",
       "6_6                   S0018c     90\n",
       "6_6                   S0018a    102\n",
       "7_7                   S0018a    102\n",
       "7_7                   S0018b     58\n",
       "7_7                   S0018c     90\n",
       "8_8                   S0018b     58\n",
       "8_8                   S0018a    102\n",
       "8_8                   S0018c     90\n",
       "9_9                S0025_28b     59\n",
       "9_9                S0025_28a     52\n",
       "9_9                   S0028b     34\n",
       "9_9                   S0028c     85\n",
       "9_9                   S0028a     36\n",
       "10_10                 S0028c     85\n",
       "10_10              S0025_28b     59\n",
       "10_10                 S0028a     36\n",
       "10_10              S0025_28a     52\n",
       "10_10                 S0028b     34\n",
       "11_11                 S0028a     36\n",
       "11_11              S0025_28b     59\n",
       "11_11                 S0028b     34\n",
       "11_11                 S0028c     85\n",
       "11_11              S0025_28a     52\n",
       "12_12                 S0017b    101\n",
       "12_12                 S0017c     55\n",
       "13_13                 S0017c     55\n",
       "13_13                 S0017b    101\n",
       "15_15                 S0022a     92\n",
       "15_15              S0021_24a     40\n",
       "15_15                 S0022c    106\n",
       "15_15              S0021_24b     86\n",
       "15_15                 S0022b     53\n",
       "16_16                 S0022c    106\n",
       "16_16                 S0022b     53\n",
       "16_16              S0021_24a     40\n",
       "16_16              S0021_24b     86\n",
       "16_16                 S0022a     92\n",
       "17_17                 S0022c    106\n",
       "17_17                 S0022a     92\n",
       "17_17                 S0022b     53\n",
       "17_17              S0021_24b     86\n",
       "17_17              S0021_24a     40\n",
       "18_18              S0021_24a     40\n",
       "18_18                 S0022c    106\n",
       "18_18                 S0022a     92\n",
       "18_18              S0021_24b     86\n",
       "18_18                 S0022b     53\n",
       "19_19              S0021_24b     86\n",
       "19_19                 S0022c    106\n",
       "19_19                 S0022b     53\n",
       "19_19              S0021_24a     40\n",
       "19_19                 S0022a     92\n",
       "23_23                 S0022a     92\n",
       "23_23              S0021_24a     40\n",
       "23_23                 S0022b     53\n",
       "23_23                 S0022c    106\n",
       "23_23              S0021_24b     86\n",
       "24_24                 S0020b     45\n",
       "24_24                 S0020c     94\n",
       "24_24                 S0020a     88\n",
       "25_25                 S0012c     71\n",
       "25_25                 S0012b      0\n",
       "25_25                 S0012a    116\n",
       "27_27                 S0012c     71\n",
       "27_27                 S0012a    116\n",
       "27_27                 S0012b      0\n",
       "28_28                 S0012c     71\n",
       "28_28                 S0012a    116\n",
       "28_28                 S0012b      0\n",
       "29_29              S0025_28a     52\n",
       "29_29                 S0026a     61\n",
       "29_29                 S0026c     56\n",
       "29_29                 S0026b     48\n",
       "29_29              S0025_28b     59\n",
       "30_30              S0025_28a     52\n",
       "30_30                 S0026a     61\n",
       "30_30                 S0026b     48\n",
       "30_30                 S0026c     56\n",
       "30_30              S0025_28b     59"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_rows\", None):\n",
    "    display(sequencing_library_mapping.sort_index().head(100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000187961</th>\n",
       "      <th>ENSG00000187583</th>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <th>ENSG00000131591</th>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <th>ENSG00000162572</th>\n",
       "      <th>ENSG00000149527</th>\n",
       "      <th>ENSG00000162591</th>\n",
       "      <th>ENSG00000238260</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000183878</th>\n",
       "      <th>ENSG00000290853</th>\n",
       "      <th>ENSG00000165246</th>\n",
       "      <th>ENSG00000176728</th>\n",
       "      <th>ENSG00000291031</th>\n",
       "      <th>ENSG00000291033</th>\n",
       "      <th>ENSG00000229236</th>\n",
       "      <th>ENSG00000198692</th>\n",
       "      <th>ENSG00000289707</th>\n",
       "      <th>ENSG00000228253</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATCGAT_S0012b</th>\n",
       "      <td>-0.098161</td>\n",
       "      <td>-0.063251</td>\n",
       "      <td>-0.087368</td>\n",
       "      <td>-0.272086</td>\n",
       "      <td>-0.240088</td>\n",
       "      <td>-0.262634</td>\n",
       "      <td>-0.068546</td>\n",
       "      <td>-0.150766</td>\n",
       "      <td>-0.213132</td>\n",
       "      <td>-0.073337</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.763635</td>\n",
       "      <td>-0.202512</td>\n",
       "      <td>-0.143501</td>\n",
       "      <td>-0.242200</td>\n",
       "      <td>-0.185859</td>\n",
       "      <td>-0.508295</td>\n",
       "      <td>-0.194552</td>\n",
       "      <td>-0.412900</td>\n",
       "      <td>-0.261451</td>\n",
       "      <td>-1.648947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAGCAACC_S0012b</th>\n",
       "      <td>-0.103190</td>\n",
       "      <td>-0.068927</td>\n",
       "      <td>-0.052226</td>\n",
       "      <td>-0.253554</td>\n",
       "      <td>-0.236903</td>\n",
       "      <td>-0.237149</td>\n",
       "      <td>-0.069453</td>\n",
       "      <td>-0.168430</td>\n",
       "      <td>-0.207583</td>\n",
       "      <td>-0.075447</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.772065</td>\n",
       "      <td>-0.184258</td>\n",
       "      <td>-0.150796</td>\n",
       "      <td>-0.208313</td>\n",
       "      <td>-0.185392</td>\n",
       "      <td>-0.486838</td>\n",
       "      <td>-0.178072</td>\n",
       "      <td>-0.334451</td>\n",
       "      <td>-0.223021</td>\n",
       "      <td>0.583485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGACAGT_S0012b</th>\n",
       "      <td>-0.115717</td>\n",
       "      <td>-0.093197</td>\n",
       "      <td>-0.121048</td>\n",
       "      <td>-0.250837</td>\n",
       "      <td>-0.193814</td>\n",
       "      <td>-0.167083</td>\n",
       "      <td>-0.073713</td>\n",
       "      <td>-0.195072</td>\n",
       "      <td>-0.187337</td>\n",
       "      <td>-0.075682</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.834996</td>\n",
       "      <td>-0.212822</td>\n",
       "      <td>-0.170287</td>\n",
       "      <td>-0.190746</td>\n",
       "      <td>-0.175803</td>\n",
       "      <td>-0.457842</td>\n",
       "      <td>-0.173275</td>\n",
       "      <td>-0.198170</td>\n",
       "      <td>-0.120289</td>\n",
       "      <td>0.288104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 3770 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ENSG00000187961  ENSG00000187583  ENSG00000188290  \\\n",
       "AAACCCAAGAATCGAT_S0012b        -0.098161        -0.063251        -0.087368   \n",
       "AAACCCAAGAGCAACC_S0012b        -0.103190        -0.068927        -0.052226   \n",
       "AAACCCAAGCGACAGT_S0012b        -0.115717        -0.093197        -0.121048   \n",
       "\n",
       "                         ENSG00000131591  ENSG00000186891  ENSG00000186827  \\\n",
       "AAACCCAAGAATCGAT_S0012b        -0.272086        -0.240088        -0.262634   \n",
       "AAACCCAAGAGCAACC_S0012b        -0.253554        -0.236903        -0.237149   \n",
       "AAACCCAAGCGACAGT_S0012b        -0.250837        -0.193814        -0.167083   \n",
       "\n",
       "                         ENSG00000162572  ENSG00000149527  ENSG00000162591  \\\n",
       "AAACCCAAGAATCGAT_S0012b        -0.068546        -0.150766        -0.213132   \n",
       "AAACCCAAGAGCAACC_S0012b        -0.069453        -0.168430        -0.207583   \n",
       "AAACCCAAGCGACAGT_S0012b        -0.073713        -0.195072        -0.187337   \n",
       "\n",
       "                         ENSG00000238260  ...  ENSG00000183878  \\\n",
       "AAACCCAAGAATCGAT_S0012b        -0.073337  ...        -0.763635   \n",
       "AAACCCAAGAGCAACC_S0012b        -0.075447  ...        -0.772065   \n",
       "AAACCCAAGCGACAGT_S0012b        -0.075682  ...        -0.834996   \n",
       "\n",
       "                         ENSG00000290853  ENSG00000165246  ENSG00000176728  \\\n",
       "AAACCCAAGAATCGAT_S0012b        -0.202512        -0.143501        -0.242200   \n",
       "AAACCCAAGAGCAACC_S0012b        -0.184258        -0.150796        -0.208313   \n",
       "AAACCCAAGCGACAGT_S0012b        -0.212822        -0.170287        -0.190746   \n",
       "\n",
       "                         ENSG00000291031  ENSG00000291033  ENSG00000229236  \\\n",
       "AAACCCAAGAATCGAT_S0012b        -0.185859        -0.508295        -0.194552   \n",
       "AAACCCAAGAGCAACC_S0012b        -0.185392        -0.486838        -0.178072   \n",
       "AAACCCAAGCGACAGT_S0012b        -0.175803        -0.457842        -0.173275   \n",
       "\n",
       "                         ENSG00000198692  ENSG00000289707  ENSG00000228253  \n",
       "AAACCCAAGAATCGAT_S0012b        -0.412900        -0.261451        -1.648947  \n",
       "AAACCCAAGAGCAACC_S0012b        -0.334451        -0.223021         0.583485  \n",
       "AAACCCAAGCGACAGT_S0012b        -0.198170        -0.120289         0.288104  \n",
       "\n",
       "[3 rows x 3770 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get the counts matrix\n",
    "cells_x_genes = adata.to_df()\n",
    "# export counts to csv\n",
    "display(cells_x_genes.head(3))\n",
    "# cells_x_genes.to_csv(f\"{outdir}/data/cells_x_genes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>onek1k_id</th>\n",
       "      <th>cpg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATCGAT_S0012b</th>\n",
       "      <td>557_558</td>\n",
       "      <td>CPG313965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAGCAACC_S0012b</th>\n",
       "      <td>27_27</td>\n",
       "      <td>CPG309211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGACAGT_S0012b</th>\n",
       "      <td>501_502</td>\n",
       "      <td>CPG313437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        onek1k_id     cpg_id\n",
       "AAACCCAAGAATCGAT_S0012b   557_558  CPG313965\n",
       "AAACCCAAGAGCAACC_S0012b     27_27  CPG309211\n",
       "AAACCCAAGCGACAGT_S0012b   501_502  CPG313437"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>geno_PC1</th>\n",
       "      <th>geno_PC2</th>\n",
       "      <th>geno_PC3</th>\n",
       "      <th>geno_PC4</th>\n",
       "      <th>geno_PC5</th>\n",
       "      <th>geno_PC6</th>\n",
       "      <th>geno_PC7</th>\n",
       "      <th>geno_PC8</th>\n",
       "      <th>geno_PC9</th>\n",
       "      <th>geno_PC10</th>\n",
       "      <th>geno_PC11</th>\n",
       "      <th>geno_PC12</th>\n",
       "      <th>geno_PC13</th>\n",
       "      <th>geno_PC14</th>\n",
       "      <th>geno_PC15</th>\n",
       "      <th>geno_PC16</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>onek1k_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>276_277</th>\n",
       "      <td>2.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.031874</td>\n",
       "      <td>-0.000410</td>\n",
       "      <td>0.004952</td>\n",
       "      <td>-0.000557</td>\n",
       "      <td>0.001924</td>\n",
       "      <td>-0.022447</td>\n",
       "      <td>0.002643</td>\n",
       "      <td>-0.001893</td>\n",
       "      <td>0.008205</td>\n",
       "      <td>-0.004277</td>\n",
       "      <td>0.003711</td>\n",
       "      <td>-0.006016</td>\n",
       "      <td>0.011691</td>\n",
       "      <td>-0.009226</td>\n",
       "      <td>0.001659</td>\n",
       "      <td>0.007041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278_279</th>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.031191</td>\n",
       "      <td>0.004006</td>\n",
       "      <td>0.003113</td>\n",
       "      <td>-0.002887</td>\n",
       "      <td>0.003075</td>\n",
       "      <td>-0.024638</td>\n",
       "      <td>0.002322</td>\n",
       "      <td>0.002359</td>\n",
       "      <td>0.021236</td>\n",
       "      <td>-0.005583</td>\n",
       "      <td>0.004636</td>\n",
       "      <td>-0.006784</td>\n",
       "      <td>0.014104</td>\n",
       "      <td>-0.007854</td>\n",
       "      <td>-0.003376</td>\n",
       "      <td>0.004050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279_280</th>\n",
       "      <td>2.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.032146</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.002034</td>\n",
       "      <td>-0.003210</td>\n",
       "      <td>-0.000880</td>\n",
       "      <td>-0.015663</td>\n",
       "      <td>0.003080</td>\n",
       "      <td>-0.001562</td>\n",
       "      <td>0.014510</td>\n",
       "      <td>-0.003109</td>\n",
       "      <td>0.004357</td>\n",
       "      <td>-0.001489</td>\n",
       "      <td>0.008275</td>\n",
       "      <td>-0.007595</td>\n",
       "      <td>-0.005692</td>\n",
       "      <td>0.002729</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sex   age  geno_PC1  geno_PC2  geno_PC3  geno_PC4  geno_PC5  \\\n",
       "onek1k_id                                                                \n",
       "276_277    2.0  65.0  0.031874 -0.000410  0.004952 -0.000557  0.001924   \n",
       "278_279    1.0  45.0  0.031191  0.004006  0.003113 -0.002887  0.003075   \n",
       "279_280    2.0  73.0  0.032146  0.001124  0.002034 -0.003210 -0.000880   \n",
       "\n",
       "           geno_PC6  geno_PC7  geno_PC8  geno_PC9  geno_PC10  geno_PC11  \\\n",
       "onek1k_id                                                                 \n",
       "276_277   -0.022447  0.002643 -0.001893  0.008205  -0.004277   0.003711   \n",
       "278_279   -0.024638  0.002322  0.002359  0.021236  -0.005583   0.004636   \n",
       "279_280   -0.015663  0.003080 -0.001562  0.014510  -0.003109   0.004357   \n",
       "\n",
       "           geno_PC12  geno_PC13  geno_PC14  geno_PC15  geno_PC16  \n",
       "onek1k_id                                                         \n",
       "276_277    -0.006016   0.011691  -0.009226   0.001659   0.007041  \n",
       "278_279    -0.006784   0.014104  -0.007854  -0.003376   0.004050  \n",
       "279_280    -0.001489   0.008275  -0.007595  -0.005692   0.002729  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample_meta = pd.read_csv(\n",
    "    \"/directflow/SCCGGroupShare/projects/anncuo/TenK10K_pilot/tenk10k/saige_qtl_input_files_covariates_sex_age_geno_pcs_tob_bioheart.csv\"\n",
    ")\n",
    "sample_meta = sample_meta[sample_meta[\"sample_id\"].isin(adata.obs[\"cpg_id\"])]\n",
    "# add in the onek1k id\n",
    "cpg_onek1k_id_mapping = adata.obs[[\"onek1k_id\", \"cpg_id\"]].drop_duplicates()\n",
    "display(cpg_onek1k_id_mapping.head(3))\n",
    "sample_meta = pd.merge(\n",
    "    sample_meta,\n",
    "    cpg_onek1k_id_mapping,\n",
    "    how=\"left\",\n",
    "    left_on=\"sample_id\",\n",
    "    right_on=\"cpg_id\",\n",
    ")\n",
    "sample_meta = sample_meta.set_index(\"onek1k_id\", drop=True).drop(\n",
    "    [\"sample_id\", \"cpg_id\"], axis=1\n",
    ") # remove these columns as I don't think the multianndata supports string variables in samplem \n",
    "display(sample_meta.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "warning: the following columns of obs are non-numeric.\n",
      "['id']\n",
      "consider casting to numeric types where appropriate, and\n",
      "consider re-coding text-valued columns with pandas.get_dummies\n"
     ]
    }
   ],
   "source": [
    "madata = mad.MultiAnnData(X=cells_x_genes, obs=cell_meta, sampleid=\"id\")\n",
    "# Add all covariate information to d.samplem\n",
    "madata.samplem = madata.samplem.join(sample_meta).join(batch_mapping)\n",
    "# check that you can actually save it with current metadata columns\n",
    "madata.write(f\"{outdir}/data/scDataObject.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2798264 × 3770\n",
       "    obs: 'id'\n",
       "    uns: 'sampleXmeta'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "madata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>geno_PC1</th>\n",
       "      <th>geno_PC2</th>\n",
       "      <th>geno_PC3</th>\n",
       "      <th>geno_PC4</th>\n",
       "      <th>geno_PC5</th>\n",
       "      <th>geno_PC6</th>\n",
       "      <th>geno_PC7</th>\n",
       "      <th>geno_PC8</th>\n",
       "      <th>geno_PC9</th>\n",
       "      <th>geno_PC10</th>\n",
       "      <th>geno_PC11</th>\n",
       "      <th>geno_PC12</th>\n",
       "      <th>geno_PC13</th>\n",
       "      <th>geno_PC14</th>\n",
       "      <th>geno_PC15</th>\n",
       "      <th>geno_PC16</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.028473</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.002113</td>\n",
       "      <td>-0.001891</td>\n",
       "      <td>0.002678</td>\n",
       "      <td>-0.012493</td>\n",
       "      <td>0.002679</td>\n",
       "      <td>-0.003106</td>\n",
       "      <td>0.010715</td>\n",
       "      <td>-0.002058</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>-0.000088</td>\n",
       "      <td>0.007350</td>\n",
       "      <td>-0.008075</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.003025</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.028473</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.002113</td>\n",
       "      <td>-0.001891</td>\n",
       "      <td>0.002678</td>\n",
       "      <td>-0.012493</td>\n",
       "      <td>0.002679</td>\n",
       "      <td>-0.003106</td>\n",
       "      <td>0.010715</td>\n",
       "      <td>-0.002058</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>-0.000088</td>\n",
       "      <td>0.007350</td>\n",
       "      <td>-0.008075</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.003025</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.028473</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.002113</td>\n",
       "      <td>-0.001891</td>\n",
       "      <td>0.002678</td>\n",
       "      <td>-0.012493</td>\n",
       "      <td>0.002679</td>\n",
       "      <td>-0.003106</td>\n",
       "      <td>0.010715</td>\n",
       "      <td>-0.002058</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>-0.000088</td>\n",
       "      <td>0.007350</td>\n",
       "      <td>-0.008075</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.003025</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.028473</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.002113</td>\n",
       "      <td>-0.001891</td>\n",
       "      <td>0.002678</td>\n",
       "      <td>-0.012493</td>\n",
       "      <td>0.002679</td>\n",
       "      <td>-0.003106</td>\n",
       "      <td>0.010715</td>\n",
       "      <td>-0.002058</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>-0.000088</td>\n",
       "      <td>0.007350</td>\n",
       "      <td>-0.008075</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.003025</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.028473</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.002113</td>\n",
       "      <td>-0.001891</td>\n",
       "      <td>0.002678</td>\n",
       "      <td>-0.012493</td>\n",
       "      <td>0.002679</td>\n",
       "      <td>-0.003106</td>\n",
       "      <td>0.010715</td>\n",
       "      <td>-0.002058</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>-0.000088</td>\n",
       "      <td>0.007350</td>\n",
       "      <td>-0.008075</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.003025</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>1.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.024420</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>0.001645</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>-0.010090</td>\n",
       "      <td>0.001389</td>\n",
       "      <td>-0.000284</td>\n",
       "      <td>0.015945</td>\n",
       "      <td>-0.006433</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>-0.001997</td>\n",
       "      <td>0.006709</td>\n",
       "      <td>-0.009851</td>\n",
       "      <td>-0.001108</td>\n",
       "      <td>0.008631</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>1.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.024420</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>0.001645</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>-0.010090</td>\n",
       "      <td>0.001389</td>\n",
       "      <td>-0.000284</td>\n",
       "      <td>0.015945</td>\n",
       "      <td>-0.006433</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>-0.001997</td>\n",
       "      <td>0.006709</td>\n",
       "      <td>-0.009851</td>\n",
       "      <td>-0.001108</td>\n",
       "      <td>0.008631</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>1.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.024420</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>0.001645</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>-0.010090</td>\n",
       "      <td>0.001389</td>\n",
       "      <td>-0.000284</td>\n",
       "      <td>0.015945</td>\n",
       "      <td>-0.006433</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>-0.001997</td>\n",
       "      <td>0.006709</td>\n",
       "      <td>-0.009851</td>\n",
       "      <td>-0.001108</td>\n",
       "      <td>0.008631</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>1.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.024420</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>0.001645</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>-0.010090</td>\n",
       "      <td>0.001389</td>\n",
       "      <td>-0.000284</td>\n",
       "      <td>0.015945</td>\n",
       "      <td>-0.006433</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>-0.001997</td>\n",
       "      <td>0.006709</td>\n",
       "      <td>-0.009851</td>\n",
       "      <td>-0.001108</td>\n",
       "      <td>0.008631</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081_1082</th>\n",
       "      <td>1.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.024420</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>0.001645</td>\n",
       "      <td>-0.000627</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>-0.010090</td>\n",
       "      <td>0.001389</td>\n",
       "      <td>-0.000284</td>\n",
       "      <td>0.015945</td>\n",
       "      <td>-0.006433</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>-0.001997</td>\n",
       "      <td>0.006709</td>\n",
       "      <td>-0.009851</td>\n",
       "      <td>-0.001108</td>\n",
       "      <td>0.008631</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3735 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sex   age  geno_PC1  geno_PC2  geno_PC3  geno_PC4  geno_PC5  \\\n",
       "id                                                                       \n",
       "1_1        2.0  65.0  0.028473  0.004421  0.002113 -0.001891  0.002678   \n",
       "1_1        2.0  65.0  0.028473  0.004421  0.002113 -0.001891  0.002678   \n",
       "1_1        2.0  65.0  0.028473  0.004421  0.002113 -0.001891  0.002678   \n",
       "1_1        2.0  65.0  0.028473  0.004421  0.002113 -0.001891  0.002678   \n",
       "1_1        2.0  65.0  0.028473  0.004421  0.002113 -0.001891  0.002678   \n",
       "...        ...   ...       ...       ...       ...       ...       ...   \n",
       "1081_1082  1.0  73.0  0.024420  0.002767  0.001645 -0.000627  0.003350   \n",
       "1081_1082  1.0  73.0  0.024420  0.002767  0.001645 -0.000627  0.003350   \n",
       "1081_1082  1.0  73.0  0.024420  0.002767  0.001645 -0.000627  0.003350   \n",
       "1081_1082  1.0  73.0  0.024420  0.002767  0.001645 -0.000627  0.003350   \n",
       "1081_1082  1.0  73.0  0.024420  0.002767  0.001645 -0.000627  0.003350   \n",
       "\n",
       "           geno_PC6  geno_PC7  geno_PC8  geno_PC9  geno_PC10  geno_PC11  \\\n",
       "id                                                                        \n",
       "1_1       -0.012493  0.002679 -0.003106  0.010715  -0.002058   0.000756   \n",
       "1_1       -0.012493  0.002679 -0.003106  0.010715  -0.002058   0.000756   \n",
       "1_1       -0.012493  0.002679 -0.003106  0.010715  -0.002058   0.000756   \n",
       "1_1       -0.012493  0.002679 -0.003106  0.010715  -0.002058   0.000756   \n",
       "1_1       -0.012493  0.002679 -0.003106  0.010715  -0.002058   0.000756   \n",
       "...             ...       ...       ...       ...        ...        ...   \n",
       "1081_1082 -0.010090  0.001389 -0.000284  0.015945  -0.006433   0.001119   \n",
       "1081_1082 -0.010090  0.001389 -0.000284  0.015945  -0.006433   0.001119   \n",
       "1081_1082 -0.010090  0.001389 -0.000284  0.015945  -0.006433   0.001119   \n",
       "1081_1082 -0.010090  0.001389 -0.000284  0.015945  -0.006433   0.001119   \n",
       "1081_1082 -0.010090  0.001389 -0.000284  0.015945  -0.006433   0.001119   \n",
       "\n",
       "           geno_PC12  geno_PC13  geno_PC14  geno_PC15  geno_PC16  batch  \n",
       "id                                                                       \n",
       "1_1        -0.000088   0.007350  -0.008075   0.001969   0.003025     48  \n",
       "1_1        -0.000088   0.007350  -0.008075   0.001969   0.003025     52  \n",
       "1_1        -0.000088   0.007350  -0.008075   0.001969   0.003025     56  \n",
       "1_1        -0.000088   0.007350  -0.008075   0.001969   0.003025     59  \n",
       "1_1        -0.000088   0.007350  -0.008075   0.001969   0.003025     61  \n",
       "...              ...        ...        ...        ...        ...    ...  \n",
       "1081_1082  -0.001997   0.006709  -0.009851  -0.001108   0.008631     38  \n",
       "1081_1082  -0.001997   0.006709  -0.009851  -0.001108   0.008631     40  \n",
       "1081_1082  -0.001997   0.006709  -0.009851  -0.001108   0.008631     57  \n",
       "1081_1082  -0.001997   0.006709  -0.009851  -0.001108   0.008631     86  \n",
       "1081_1082  -0.001997   0.006709  -0.009851  -0.001108   0.008631     91  \n",
       "\n",
       "[3735 rows x 19 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "madata.samplem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>geno_PC1</th>\n",
       "      <th>geno_PC2</th>\n",
       "      <th>geno_PC3</th>\n",
       "      <th>geno_PC4</th>\n",
       "      <th>geno_PC5</th>\n",
       "      <th>geno_PC6</th>\n",
       "      <th>geno_PC7</th>\n",
       "      <th>geno_PC8</th>\n",
       "      <th>geno_PC9</th>\n",
       "      <th>geno_PC10</th>\n",
       "      <th>geno_PC11</th>\n",
       "      <th>geno_PC12</th>\n",
       "      <th>geno_PC13</th>\n",
       "      <th>geno_PC14</th>\n",
       "      <th>geno_PC15</th>\n",
       "      <th>geno_PC16</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>onek1k_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>276_277</th>\n",
       "      <td>2.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.031874</td>\n",
       "      <td>-0.000410</td>\n",
       "      <td>0.004952</td>\n",
       "      <td>-0.000557</td>\n",
       "      <td>0.001924</td>\n",
       "      <td>-0.022447</td>\n",
       "      <td>0.002643</td>\n",
       "      <td>-0.001893</td>\n",
       "      <td>0.008205</td>\n",
       "      <td>-0.004277</td>\n",
       "      <td>0.003711</td>\n",
       "      <td>-0.006016</td>\n",
       "      <td>0.011691</td>\n",
       "      <td>-0.009226</td>\n",
       "      <td>0.001659</td>\n",
       "      <td>0.007041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278_279</th>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.031191</td>\n",
       "      <td>0.004006</td>\n",
       "      <td>0.003113</td>\n",
       "      <td>-0.002887</td>\n",
       "      <td>0.003075</td>\n",
       "      <td>-0.024638</td>\n",
       "      <td>0.002322</td>\n",
       "      <td>0.002359</td>\n",
       "      <td>0.021236</td>\n",
       "      <td>-0.005583</td>\n",
       "      <td>0.004636</td>\n",
       "      <td>-0.006784</td>\n",
       "      <td>0.014104</td>\n",
       "      <td>-0.007854</td>\n",
       "      <td>-0.003376</td>\n",
       "      <td>0.004050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279_280</th>\n",
       "      <td>2.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.032146</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.002034</td>\n",
       "      <td>-0.003210</td>\n",
       "      <td>-0.000880</td>\n",
       "      <td>-0.015663</td>\n",
       "      <td>0.003080</td>\n",
       "      <td>-0.001562</td>\n",
       "      <td>0.014510</td>\n",
       "      <td>-0.003109</td>\n",
       "      <td>0.004357</td>\n",
       "      <td>-0.001489</td>\n",
       "      <td>0.008275</td>\n",
       "      <td>-0.007595</td>\n",
       "      <td>-0.005692</td>\n",
       "      <td>0.002729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280_281</th>\n",
       "      <td>2.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.022925</td>\n",
       "      <td>-0.001849</td>\n",
       "      <td>0.008604</td>\n",
       "      <td>-0.003138</td>\n",
       "      <td>0.001227</td>\n",
       "      <td>-0.012471</td>\n",
       "      <td>-0.000189</td>\n",
       "      <td>-0.000706</td>\n",
       "      <td>0.019296</td>\n",
       "      <td>-0.005162</td>\n",
       "      <td>0.004675</td>\n",
       "      <td>-0.011633</td>\n",
       "      <td>0.001585</td>\n",
       "      <td>-0.012458</td>\n",
       "      <td>0.001691</td>\n",
       "      <td>0.002326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281_282</th>\n",
       "      <td>2.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.029691</td>\n",
       "      <td>0.004519</td>\n",
       "      <td>0.004947</td>\n",
       "      <td>-0.002352</td>\n",
       "      <td>0.002671</td>\n",
       "      <td>-0.021663</td>\n",
       "      <td>0.005412</td>\n",
       "      <td>-0.002743</td>\n",
       "      <td>0.019496</td>\n",
       "      <td>-0.011135</td>\n",
       "      <td>0.008282</td>\n",
       "      <td>-0.007981</td>\n",
       "      <td>0.013814</td>\n",
       "      <td>-0.013639</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.009120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959_960</th>\n",
       "      <td>2.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.030631</td>\n",
       "      <td>0.003893</td>\n",
       "      <td>0.004568</td>\n",
       "      <td>-0.002958</td>\n",
       "      <td>0.001628</td>\n",
       "      <td>-0.020637</td>\n",
       "      <td>0.003194</td>\n",
       "      <td>-0.002507</td>\n",
       "      <td>0.025744</td>\n",
       "      <td>-0.007174</td>\n",
       "      <td>0.001096</td>\n",
       "      <td>-0.013423</td>\n",
       "      <td>0.013252</td>\n",
       "      <td>-0.015006</td>\n",
       "      <td>-0.001747</td>\n",
       "      <td>0.010526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960_961</th>\n",
       "      <td>2.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.029611</td>\n",
       "      <td>0.001636</td>\n",
       "      <td>0.002385</td>\n",
       "      <td>-0.001334</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>-0.015766</td>\n",
       "      <td>0.002925</td>\n",
       "      <td>0.001546</td>\n",
       "      <td>0.023328</td>\n",
       "      <td>-0.002993</td>\n",
       "      <td>0.004423</td>\n",
       "      <td>-0.005617</td>\n",
       "      <td>0.009979</td>\n",
       "      <td>-0.014310</td>\n",
       "      <td>0.001826</td>\n",
       "      <td>0.009812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>961_962</th>\n",
       "      <td>2.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>0.027068</td>\n",
       "      <td>0.004368</td>\n",
       "      <td>0.005220</td>\n",
       "      <td>-0.003066</td>\n",
       "      <td>0.002257</td>\n",
       "      <td>-0.016874</td>\n",
       "      <td>0.000897</td>\n",
       "      <td>-0.002162</td>\n",
       "      <td>0.018960</td>\n",
       "      <td>-0.006821</td>\n",
       "      <td>0.003939</td>\n",
       "      <td>-0.007990</td>\n",
       "      <td>0.009144</td>\n",
       "      <td>-0.013954</td>\n",
       "      <td>-0.006080</td>\n",
       "      <td>0.001970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962_963</th>\n",
       "      <td>1.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.032541</td>\n",
       "      <td>0.001841</td>\n",
       "      <td>0.005458</td>\n",
       "      <td>-0.003339</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>-0.019975</td>\n",
       "      <td>0.004457</td>\n",
       "      <td>-0.000195</td>\n",
       "      <td>0.019118</td>\n",
       "      <td>-0.001334</td>\n",
       "      <td>0.004132</td>\n",
       "      <td>-0.007752</td>\n",
       "      <td>0.006904</td>\n",
       "      <td>-0.010907</td>\n",
       "      <td>0.004262</td>\n",
       "      <td>0.003227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685_686</th>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.024478</td>\n",
       "      <td>0.002551</td>\n",
       "      <td>-0.000728</td>\n",
       "      <td>-0.000339</td>\n",
       "      <td>0.000727</td>\n",
       "      <td>-0.015689</td>\n",
       "      <td>-0.000067</td>\n",
       "      <td>0.000745</td>\n",
       "      <td>0.023395</td>\n",
       "      <td>-0.004391</td>\n",
       "      <td>0.000521</td>\n",
       "      <td>-0.007543</td>\n",
       "      <td>0.017150</td>\n",
       "      <td>-0.014151</td>\n",
       "      <td>-0.001830</td>\n",
       "      <td>0.000962</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>976 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sex   age  geno_PC1  geno_PC2  geno_PC3  geno_PC4  geno_PC5  \\\n",
       "onek1k_id                                                                \n",
       "276_277    2.0  65.0  0.031874 -0.000410  0.004952 -0.000557  0.001924   \n",
       "278_279    1.0  45.0  0.031191  0.004006  0.003113 -0.002887  0.003075   \n",
       "279_280    2.0  73.0  0.032146  0.001124  0.002034 -0.003210 -0.000880   \n",
       "280_281    2.0  25.0  0.022925 -0.001849  0.008604 -0.003138  0.001227   \n",
       "281_282    2.0  53.0  0.029691  0.004519  0.004947 -0.002352  0.002671   \n",
       "...        ...   ...       ...       ...       ...       ...       ...   \n",
       "959_960    2.0  19.0  0.030631  0.003893  0.004568 -0.002958  0.001628   \n",
       "960_961    2.0  73.0  0.029611  0.001636  0.002385 -0.001334  0.000352   \n",
       "961_962    2.0  62.0  0.027068  0.004368  0.005220 -0.003066  0.002257   \n",
       "962_963    1.0  57.0  0.032541  0.001841  0.005458 -0.003339  0.001115   \n",
       "685_686    NaN  75.0  0.024478  0.002551 -0.000728 -0.000339  0.000727   \n",
       "\n",
       "           geno_PC6  geno_PC7  geno_PC8  geno_PC9  geno_PC10  geno_PC11  \\\n",
       "onek1k_id                                                                 \n",
       "276_277   -0.022447  0.002643 -0.001893  0.008205  -0.004277   0.003711   \n",
       "278_279   -0.024638  0.002322  0.002359  0.021236  -0.005583   0.004636   \n",
       "279_280   -0.015663  0.003080 -0.001562  0.014510  -0.003109   0.004357   \n",
       "280_281   -0.012471 -0.000189 -0.000706  0.019296  -0.005162   0.004675   \n",
       "281_282   -0.021663  0.005412 -0.002743  0.019496  -0.011135   0.008282   \n",
       "...             ...       ...       ...       ...        ...        ...   \n",
       "959_960   -0.020637  0.003194 -0.002507  0.025744  -0.007174   0.001096   \n",
       "960_961   -0.015766  0.002925  0.001546  0.023328  -0.002993   0.004423   \n",
       "961_962   -0.016874  0.000897 -0.002162  0.018960  -0.006821   0.003939   \n",
       "962_963   -0.019975  0.004457 -0.000195  0.019118  -0.001334   0.004132   \n",
       "685_686   -0.015689 -0.000067  0.000745  0.023395  -0.004391   0.000521   \n",
       "\n",
       "           geno_PC12  geno_PC13  geno_PC14  geno_PC15  geno_PC16  \n",
       "onek1k_id                                                         \n",
       "276_277    -0.006016   0.011691  -0.009226   0.001659   0.007041  \n",
       "278_279    -0.006784   0.014104  -0.007854  -0.003376   0.004050  \n",
       "279_280    -0.001489   0.008275  -0.007595  -0.005692   0.002729  \n",
       "280_281    -0.011633   0.001585  -0.012458   0.001691   0.002326  \n",
       "281_282    -0.007981   0.013814  -0.013639   0.000209   0.009120  \n",
       "...              ...        ...        ...        ...        ...  \n",
       "959_960    -0.013423   0.013252  -0.015006  -0.001747   0.010526  \n",
       "960_961    -0.005617   0.009979  -0.014310   0.001826   0.009812  \n",
       "961_962    -0.007990   0.009144  -0.013954  -0.006080   0.001970  \n",
       "962_963    -0.007752   0.006904  -0.010907   0.004262   0.003227  \n",
       "685_686    -0.007543   0.017150  -0.014151  -0.001830   0.000962  \n",
       "\n",
       "[976 rows x 18 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data vis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize data\n",
    "# plt.scatter(madata.obsm['X_umap'][d.obs.clust1==1,0], d.obsm['X_umap'][d.obs.clust1==1,1],\n",
    "#             c=\"green\", edgecolor='none', s=2, label = \"Cluster 1\")\n",
    "# plt.scatter(madata.obsm['X_umap'][d.obs.clust2==1,0], d.obsm['X_umap'][d.obs.clust2==1,1],\n",
    "#             c=\"purple\", edgecolor='none', s=2, label = \"Cluster 2\")\n",
    "# plt.scatter(madata.obsm['X_umap'][d.obs.clust3==1,0], d.obsm['X_umap'][d.obs.clust3==1,1],\n",
    "#             c=\"orange\", edgecolor='none', s=2, label = \"Cluster 3\")\n",
    "# plt.legend(loc=\"lower left\", markerscale=7, frameon=False)\n",
    "# plt.axis(\"off\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other required inputs:\n",
    "* plink2 format genotyping data for each sample"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
